### ISSUE TITLE
Incompatible shapes

### ISSUE THREAD

**Issue by faresobeid** [jonmorton/gateloop_jax#1]
I think there is a mistake with the line (k @ v).astype(...), as there was no where before it where k or v were transposed. The only way this could have gone without an error if the hidden dimension is equal to the sequence length. Could you please clarify which of those to transpose and fix the issue. Thanks

**Comment by MasahideOkada** [comment]
I think `k * v` is correct here if the shape of `k` and `v` is supposed to be `(batch_size, seq_len, hidden_dim)`, which is described as *max-head* in the paper.
To my understanding, matrix multiplication `k @ v` is not appropriate because `kv` is, generally, the outer product of `k` and `v` according to the paper.
The author said the shape of `kv` is `(batch_size, seq_len, num_heads, hidden_dim, hidden_dim)` in [this discussion](https://github.com/lucidrains/gateloop-transformer/discussions/1#discussioncomment-7516938).
Therefore, the shape of `k` and `v` should be `(batch_size, seq_len, num_heads, hidden_dim)`.
When `hidden_dim == 1`, it's max-head and the outer product of `k` and `v` equals to their element-wise product.
We can check it from this code.

```py
import jax
import jax.numpy as jnp
from jax import random

@jax.jit
def outer_product(x, y):
    return jnp.einsum("...i, ...j -> ...ij", x, y)

key = random.key(0)
k1, k2 = random.split(key)
batch_size = 5
seq_len = 32
num_heads = 4
hidden_dim = 1
shape = (batch_size, seq_len, num_heads, hidden_dim)
k = random.normal(k1, shape)
v = random.normal(k2, shape)
kv_outer = outer_product(k, v) # outer product
kv_elemwise = k * v # element-wise product
print(jnp.allclose(kv_outer.squeeze(axis=-1), kv_elemwise)) # will print True
```

For the max-head gate loop, I implemented it as

```py
import jax
import jax.numpy as jnp
from jax import random
from jax.lax import associative_scan

# max-headed gate loop
@jax.jit
def max_headed_gate_loop_operator(k, v, q, a):
    def binary_operator(e_i, e_j):
        a_i, kv_i = e_i
        a_j, kv_j = e_j
        return a_j * a_i, a_j * kv_i + kv_j

    kv = (k * v).astype(a.dtype)
    _, y = associative_scan(binary_operator, (a, kv), axis=1)
    y = q.astype(y.dtype) * y
    return y

key = random.key(0)
keys = random.split(key, num=5)

batch_size = 16
seq_len = 32
dim = 64
shape = (batch_size, seq_len, dim)

q = random.normal(keys[0], shape)
k = random.normal(keys[1], shape)
v = random.normal(keys[2], shape)
a = jax.lax.complex(
    random.normal(keys[3], shape),
    random.normal(keys[4], shape)
)
y1 = max_headed_gate_loop_operator(k, v, q, a)
```

and I attempted to make the general gate loop like

```py
# general gate loop
@jax.jit
def matmul1(x, y):
    return jnp.einsum("...i, ...ik -> ...k", x, y)

@jax.jit
def matmul2(x, y):
    return jnp.einsum("...i, ...ik -> ...ik", x, y)

@jax.jit
def general_gate_loop_operator(k, v, q, a):
    def binary_operator(e_i, e_j):
        a_i, kv_i = e_i
        a_j, kv_j = e_j
        return a_j * a_i, matmul2(a_j, kv_i) + kv_j

    kv = outer_product(k, v).astype(a.dtype)
    _, y = associative_scan(binary_operator, (a, kv), axis=1)
    y = matmul1(q.astype(y.dtype), y)
    return y

key = random.key(0)
keys = random.split(key, num=5)

batch_size = 16
seq_len = 32
num_heads = 64
dim = 1
shape = (batch_size, seq_len, num_heads, dim)

q = random.normal(keys[0], shape)
k = random.normal(keys[1], shape)
v = random.normal(keys[2], shape)
a = jax.lax.complex(
    random.normal(keys[3], shape),
    random.normal(keys[4], shape),
)
y2 = general_gate_loop_operator(k, v, q, a)
```

We set `dim = 1` for the general one, so it's max-head and `y1` and `y2` must be the same.

```py
print(jnp.allclose(y2.squeeze(axis=-1), y1)) # will print True
```
